tools/opt/opt.cpp

```diff
@@ -505,182 +500,53 @@ OptStatus ParseFlags(int argc, const char** argv, Optimizer* optimizer,
-        if (status.action != OPT_CONTINUE) {
-          return status;
-        }
-      } else if (0 == strcmp(cur_arg, "--loop-unroll")) {
-        optimizer->RegisterPass(CreateLoopUnrollPass(true));
-      } else if (0 == strcmp(cur_arg, "--vector-dce")) {
-        optimizer->RegisterPass(CreateVectorDCEPass());
-      } else if (0 == strcmp(cur_arg, "--loop-unroll-partial")) {
-        OptStatus status =
-            ParseLoopUnrollPartialArg(argc, argv, ++argi, optimizer);
-        if (status.action != OPT_CONTINUE) {
-          return status;
-        }
-      } else if (0 == strcmp(cur_arg, "--loop-peeling")) {
-        optimizer->RegisterPass(CreateLoopPeelingPass());
-      } else if (0 == strcmp(cur_arg, "--loop-peeling-threshold")) {
-        OptStatus status = ParseLoopPeelingThresholdArg(argc, argv, ++argi);
-        if (status.action != OPT_CONTINUE) {
-          return status;
-        }
-      } else if (0 == strcmp(cur_arg, "--skip-validation")) {
-        *skip_validator = true;
-      } else if (0 == strcmp(cur_arg, "-O")) {
-        optimizer->RegisterPerformancePasses();
-      } else if (0 == strcmp(cur_arg, "-Os")) {
-        optimizer->RegisterSizePasses();
-      } else if (0 == strcmp(cur_arg, "--legalize-hlsl")) {
-        *skip_validator = true;
-        optimizer->RegisterLegalizationPasses();
       } else if (0 == strncmp(cur_arg, "-Oconfig=", sizeof("-Oconfig=") - 1)) {
         OptStatus status =
             ParseOconfigFlag(argv[0], cur_arg, optimizer, in_file, out_file);
         if (status.action != OPT_CONTINUE) {
           return status;
         }
-      } else if (0 == strcmp(cur_arg, "--ccp")) {
-        optimizer->RegisterPass(CreateCCPPass());
+      } else if (0 == strcmp(cur_arg, "--skip-validation")) {
+        *skip_validator = true;
       } else if (0 == strcmp(cur_arg, "--print-all")) {
         optimizer->SetPrintAll(&std::cerr);
       } else if (0 == strcmp(cur_arg, "--time-report")) {
         optimizer->SetTimeReport(&std::cerr);
-      } else if ('\0' == cur_arg[1]) {
-        // Setting a filename of "-" to indicate stdin.
-        if (!*in_file) {
-          *in_file = cur_arg;
-        } else {
-          fprintf(stderr, "error: More than one input file specified\n");
-          return {OPT_STOP, 1};
-        }
+      } else if (0 == strcmp(cur_arg, "--relax-struct-store")) {
+        options->relax_struct_store = true;
       } else {
-        fprintf(
-            stderr,
-            "error: Unknown flag '%s'. Use --help for a list of valid flags\n",
-            cur_arg);
-        return {OPT_STOP, 1};
+        // Some passes used to accept the form '--pass arg', canonicalize them
+        // to '--pass=arg'.
+        pass_flags.push_back(CanonicalizeFlag(argv, argc, &argi));
+
+        // If we were requested to legalize SPIR-V generated from the HLSL
+        // front-end, skip validation.
+        if (0 == strcmp(cur_arg, "--legalize-hlsl")) *skip_validator = true;
```

Comment:
This seems to be in a weird place. Also, this is not done if --legalize-hlsl is used in -Oconfig.

Reply:
It needs to be there, because the flag affects the behaviour of spirv-opt.  The flag will be done out of -Oconfig because the processing of Oconfig flags simply reads the strings and passes them out to ParseFlags() recursively.

So, it doesn't matter where flags are read from.